base/io_buf.h

```diff
@@ -44,85 +56,61 @@ class IoBuf {
     return Bytes{buf_ + offs_, InputLen()};
   }
 
-  size_t InputLen() const {
-    return size_ - offs_;
-  }
+  // Mark num_read bytes from the input as read.
+  void ConsumeInput(size_t num_read);
+
+  // Write num_write bytes to dest and mark them as read.
+  void ReadAndConsume(size_t num_write, void* dest);
 
-  void ConsumeInput(size_t offs);
+  // ============== OUTPUT ============
 
-  void ReadAndConsume(size_t sz, void* dest) {
-    ConstBytes b = InputBuffer();
-    assert(b.size() >= sz);
-    memcpy(dest, b.data(), sz);
-    ConsumeInput(sz);
+  size_t AppendLen() {
+    return capacity_ - size_;
   }
 
   Bytes AppendBuffer() {
-    return Bytes{buf_ + size_, Available()};
+    return Bytes{buf_ + size_, AppendLen()};
   }
 
-  void CommitWrite(size_t sz) {
-    size_ += sz;
+  // Mark num_written bytes as written and transform them to input.
+  void CommitWrite(size_t num_written) {
+    size_ += num_written;
   }
 
-  void Clear() {
-    size_ = 0;
-    offs_ = 0;
-  }
+  // Copy num_copy bytes from source to append buffer and mark them as written.
+  // Ensures append buffer is large enough.
+  void WriteAndCommit(size_t num_copy, const void* source);
 
+  // Ensure required append buffer size.
   void EnsureCapacity(size_t sz) {
-    Reserve(InputLen() + sz);
+    Reserve(size_ + sz);
   }
 
-  void Reserve(size_t sz) {
-    if (sz < capacity_)
-      return;
-
-    sz = absl::bit_ceil(sz);
-    uint8_t* nb = new (std::align_val_t{alignment_}) uint8_t[sz];
-    if (buf_) {
-      if (size_ > offs_) {
-        memcpy(nb, buf_ + offs_, size_ - offs_);
-        size_ -= offs_;
-        offs_ = 0;
-      } else {
-        size_ = offs_ = 0;
-      }
-      delete[] buf_;
-    }
-
-    buf_ = nb;
-    capacity_ = sz;
-  }
+  // Reserve by whole buffer size.
+  // Use EnsureCapacity instead for resizing only by desired append buffer size.
+  void Reserve(size_t full_size);
```

Comment:
Those function is dangerously misleading, it accounts only for the TOTAL size. Most of the time EnsureCapacity is really needed from outside accessors

Reply:
if you think it's not needed, feel free to remove it...